Updated uq interpreter

utting opened the pull request:
An interpreter for GraalVM IR graphs.  
This can be useful for evaluating IR graphs during testing.

This implements 'interpret(InterpreterState)' methods in 30+ of the most common subclasses of FixedNode in order to interpret those control-flow nodes, and 'interpretExpr(InterpreterState)' methods in most of the common data flow nodes in order to interpret expressions.  The main entry point for the interpreter is the GraalInterpreter class and its executeGraph method.

The interpreter is based on a 2020-2021 UQ honours project by Leon Karstens plus further contributions by Kristian Thomassen and Mark Utting (UQ).  Some current limitations are that it focusses mostly on high-level nodes rather than lowered nodes, supports only simple objects whose fields are primitive values, it attempts to interpret methods in ALL classes that are called (an alternative might be to treat some java.lang.* classes as pre-defined), and uses reflection to initialize static fields (which means that it cannot find some private fields).

kethomassen committed:
Move over interpreter code from uq_interpreter branch

kethomassen committed:
Run eclipseformat

utting committed:
Better error messages for binary arithmetic evaluations that go wrong in the interpreter.

utting committed:
Ignore the interpreter result for unit tests that use nodes that are not implemented yet.

utting committed:
Fix interpreter bug when initializing static fields - was not using the correct value()

utting committed:
Better error messages when the interpreter fails.  Initialize ALL static fields including private ones.

utting committed:
Fix infinite loop in interpreter evaluation of LoadIndexedNode

utting committed:
Improve interpreter to initialize more static fields, but skip those with java.lang.reflect.InaccessibleObjectException issues.

utting committed:
Rename interpretControlFlow(_) method to interpret(_) and interpretDataFlow(_) to interpretExpr(_).  And FixedNode provides a useful default implementation of the latter.

utting committed:
Add javadocs.  Move visitMerge from interpreter into AbstractMergeNode.  Make Node lookups total.  Tag thrown exceptions with an unwinding flag to distinguish them from Exception objects returned normally.

utting committed:
Fixed phi node updates to be all done in parallel.

utting committed:
Adding verbose tracing to interpreter execution, and experiment with supporting constant real Java objects.

utting committed:
Resolve import differences between updated-master and updated-interpreter

utting committed:
Fix Checkstyle issues.

utting committed:
Merge pull request #38 from uqcyber/updated-interpreter

Add IR Interpreter, with support for simple objects.

utting committed:
Fix docs and unused variable.

utting committed:
Renaming to fix some minor issues found by gate tests.

utting committed:
Merge branch 'oracle:master' into updated-master

BraeWebb committed:
Resolve style violations from mx gate

dougxc reviewed (commented):
I have done a cursory review of this PR and left a few comments.
While it's a good start, it will need substantial work before it might be integrated. Namely, all the TODO's should be addressed and other unimplemented functionality should be completed.

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

please note that I delete InterperterValueConstantObject.java in my personal test environment

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Added some test cases for objects.

WeiyiDong committed:
Removed graal.Interpreter flag testing

Removed graal.Interpreter flag testing to enable interpreter test temporarily.

WeiyiDong committed:
Fixed activation stack. Added JVMContext

Fixed activation stack. Added JVMContext to the GraalInterpreter constructor and some other constructors. Added JVMContextImpl class. Adapted InterpreterValueFactoryImpl to native object impl of InterpreterValueMutableObject. Removed interpreter heap.

WeiyiDong committed:
Added a getTypeClass() method 

Added a getTypeClass() method used by InterpreterValueMutableObject and InterpreterValueArray

WeiyiDong committed:
Support primitive array.

Support primitive array. When accessing array elements, convert elements from native objects into InterpreterValue back and forth.

WeiyiDong committed:
Use native objects

Use native objects to implement interpreter objects. Use reflection to construct native object, and VarHandles to access fields.

WeiyiDong committed:
same with the orginal format

same with the orginal format

WeiyiDong committed:
Added some static methods

Added some static methods to convert primitive values into InterpreterValuePrimitive

WeiyiDong committed:
support cross-module

Added this interface to support cross-module class loading and VarHandle use for InterpreterValueMutableObject.

WeiyiDong committed:
Support null constant.

Support null constant.

WeiyiDong committed:
Support null constant.

Support null constant.

WeiyiDong committed:
Support null constant.

Support null constant.

WeiyiDong committed:
Removed getHeapValue() and setHeapValue().

Removed getHeapValue() and setHeapValue().

WeiyiDong committed:
distinguish exception 

Added this exception to distinguish interpretation exception and exception caused by interpreter malfunctioning.

WeiyiDong committed:
Throws exception 

Throws exception if the object is null in interpreter

WeiyiDong committed:
Throws exception

Throws exception if the array is null in interpreter

WeiyiDong committed:
setHeapValue() -> setNodeLookupValue()

setHeapValue() -> setNodeLookupValue()

WeiyiDong committed:
setHeapValue() -> setNodeLookupValue()

setHeapValue() -> setNodeLookupValue()

WeiyiDong committed:
setHeapValue() -> setNodeLookupValue()

setHeapValue() -> setNodeLookupValue()

WeiyiDong committed:
Throws exception 

Throws exception if the object is null in interpreter

WeiyiDong committed:
Throws exception 

Throws exception if the array is null in interpreter

WeiyiDong committed:
Fixed object equality testing in interpreter.

Fixed object equality testing in interpreter.

WeiyiDong committed:
using Unsafe to create object

WeiyiDong committed:
add dependency

WeiyiDong committed:
use Unsafe to create object

WeiyiDong committed:
use Unsafe to create object

WeiyiDong committed:
add constructor tests

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
support cross-moudle

WeiyiDong committed:
Add files via upload

WeiyiDong committed:
Fixed object equality testing in interpreter.

WeiyiDong committed:
Throws exception

WeiyiDong committed:
distinguish exception
